### DISCUSSION TITLE
module installations for perl

### DISCUSSION THREAD

**Discussion by laurel-nave-powers** [sr320/course-fish546-2021#68] [Q&A] [answered]
I am running into an issue installing a module to be able to run a perl script to trim adaptors. I have the data in .fq format ready to go. Right now I just need to get the Parallel/Fork Manager installed. I dragged and dropped the folder that was downloaded for the module into the location where my perl is running from (that I found from the error message, see the jupyter notebook below) and it isn't compressed. But it still isn't working. I would appreciate any insight anyone might have. 
This is the Jupyter notebook I am working in for this : https://github.com/fish546-2021/Laurel-genes/blob/main/trim.ipynb 
And this is the perl script I am trying to use: https://github.com/fish546-2021/Laurel-genes/blob/main/trim_adaptor.pl 
Thank you!

**Comment by kubu4** [comment]
The error message indicates that it couldn't find that file - which is correct.

Your Perl installation is in a different directory than where you were in the Jupyter Notebook when you tried to execute that installation script.

Related note, I'd highly recommend avoiding running software installations from a Jupyter Notebook. Instead, do so directly from command line.

**Reply by laurel-nave-powers** [reply]
Thank you for your response! 
So you are saying my Jupyter Notebook and/or the perl script need to be in the same folder as the perl and module installation?

**Comment by kubu4** [comment]
Ignore the Jupyter Notebook for now; don't use it for this.

From Terminal, you need to navigate to the directory that has the Perl file you're trying to execute (Makefile. PL) and then run the code to perform the installation.

**Reply by laurel-nave-powers** [reply]
Ok, I think I have all the many modules correctly downloaded. I also had to download two packages, "trim_galore" and "cutadapt". Now when I try and run the script it says that trim_galore can't be found and is not in the $PATH. I have gone in and added the location everything is in to my path using "sudo nano /etc/paths" but when I go to check it using "echo $PATH" it shows up as the same as before. 
Not sure what I am doing wrong and getting confused. 
Thank you for any help!

**Comment by kubu4** [comment, marked as answer]
For appending locations to your system `$PATH`, you generally want to update your `~/.zshrc` file. The location you're currently using (`/etc/paths`) is a non-standard location/file that your system won't automatically look at for information. The `~/.zsrhrc` file, however, is part of the default set of files that the shell will look at to gather information when staring.

So, you'd want to do something like this:

1. Open your `~/.zshrc` file for editing:

    `nano ~/.zshrc`

2. Add something like this to the _end_ of the file (add both lines). For me, it would look like this (Also, be sure to change the path below to match the path to the _folder_ where the trimgalore program is on your computer.):

```bash
# Add to PATH
 export PATH=${PATH}:/home/sam/.local/bin
```

3. Exit nano and save changes.

4. Close Terminal (close the entire program, not just a single window).

5. Open Terminal and confirm your system `$PATH` is updated:

`echo $PATH`

If things worked, then you'll see something like `:/home/sam/.local/bin` at the end of the output from the command (yours should reflect the trimgalore folder path you specified in `#2` above.

Repeat for the cutadapt folder.

**Reply by laurel-nave-powers** [reply]
Thank you! That seems to have worked for trim_galore. I tried to do the same thing for cutadapt but it is giving the same error. It is in the same folder as trim_galore and I can see it when I list out the contents. I added the same PATH again just to make sure I didn't need it twice, but I get the same error. I tried moving cutadapt into its own folder within the folder and added that PATH but this didn't work either (in fact it made it so I couldn't open jupyter-lab until I deleted that particular PATH).

**Comment by kubu4** [comment]
If possible, please add screenshots with some of these troubleshooting things. If possible, make sure screenshots include results from `pwd` command, results from `echo $PATH`, and, if the results aren't too long, `ls` of the directories where programs are installed.

How did you install cutadapt? There are multiple ways to do so and they have different implications on how to add it to the system $PATH.

Finally, this may be easier to address with some live help during class today.

**Reply by laurel-nave-powers** [reply]
Here is my updated notebook that has pwd, echo $path, and ls at the bottom (https://github.com/fish546-2021/Laurel-genes/blob/main/trim.ipynb). 
The process for installing cutadapt was complicated. First I had to set up channels and install bioconda (https://bioconda.github.io/user/install.html). Then I had to install cutadapt into a new conda environment (https://cutadapt.readthedocs.io/en/stable/installation.html). 
It says an environment needs to be activated every time I want to use the programs in it. So I tried that using `conda activate cutadaptenv` but that hasn't seemed to help. 

Yes, live help will probably be easier. 
Thank you for all your help

**Comment by kubu4** [comment]
Did you try to follow this suggestion from that last error message?

```
CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
To initialize your shell, run
conda init <SHELL_NAME>
```

In Terminal, you would run:

`conda init zsh`

Close Terminal entirely (and shutdown any running Jupyter Notebooks), open Terminal and then try to activate your cutadapt environment again and see what happens.

**Reply by laurel-nave-powers** [reply]
Yes, I did something similar and it worked in my terminal. Now other errors... yay.
